lib/internal/modules/esm/loader.js

```diff
@@ -408,41 +481,103 @@ class ESMLoader {
-      namespaces[0];
+    if (!wasArr) { return namespaces[0]; }
+
+    for (let i = 0; i < count; i++) {
+      const namespace = ObjectCreate(null);
+      namespace.url = specifiers[i];
+      namespace.exports = namespaces[i];
+
+      namespaces[i] = namespace;
+    }
+
+    return namespaces;
   }
 
   /**
    * Provide source that is understood by one of Node's translators.
    *
-   * The internals of this WILL change when chaining is implemented,
-   * depending on the resolution/consensus from #36954
-   * @param {string} url The URL/path of the module to be loaded
+   * Internally, this behaves like a backwards iterator, wherein the stack of
+   * hooks starts at the top and each call to `next()` moves down 1 step until
+   * it reaches the bottom or short-circuits.
+   *
+   * @param {URL['href']} url The URL/path of the module to be loaded
    * @param {object} context Metadata about the module
-   * @returns {object}
+   * @returns {{ format: ModuleFormat, source: ModuleSource }}
    */
   async load(url, context = {}) {
-    const defaultLoader = this.#loaders[0];
+    const loaders = this.#loaders;
+    let hookIndex = loaders.length - 1;
+    let {
+      fn: loader,
+      url: loaderFilePath,
+    } = loaders[hookIndex];
+    let chainFinished = hookIndex === 0;
+
+    const next = (nextUrl, ctx = context) => {
+      // `next` has been called, so decrement our pointer
+      --hookIndex;
+
+      ({
+        fn: loader,
+        url: loaderFilePath,
+      } = loaders[hookIndex]);
+
+      if (hookIndex === 0) { chainFinished = true; }
+
+      const hookErrIdentifier = `${loaderFilePath} "load"`;
+
+      if (typeof nextUrl !== 'string') {
+        // non-strings can be coerced to a url string
+        // validateString() throws a less-specific error
+        throw new ERR_INVALID_ARG_TYPE(
+          `${hookErrIdentifier} next(url)`,
+          'a url string',
+          nextUrl,
+        );
+      }
 
-    const loader = this.#loaders.length === 1 ?
-      defaultLoader :
-      this.#loaders[1];
-    const loaded = await loader(url, context, defaultLoader);
+      try {
+        new URL(nextUrl);
+      } catch {
+        throw new ERR_INVALID_ARG_VALUE(
+          `${hookErrIdentifier} next(url)`,
+          nextUrl,
+          'should be a url string',
+        );
+      }
+
+      validateObject(ctx, `${hookErrIdentifier} next(, context)`);
 
-    if (typeof loaded !== 'object') {
+      return loader(nextUrl, ctx, next);
+    };
+
+    const loaded = await loader(
+      url,
+      context,
+      next,
+    );
+
+    const hookErrIdentifier = `${loaderFilePath} load`;
+
+    if (typeof loaded !== 'object') { // [2]
       throw new ERR_INVALID_RETURN_VALUE(
-        'object',
-        'loader load',
+        'an object',
+        hookErrIdentifier,
         loaded,
       );
     }
 
     const {
       format,
+      shortCircuit,
       source,
     } = loaded;
 
+    if (!chainFinished && !shortCircuit) {
```

Comment:
As discussed separately with @JakobJingleheimer I think it would be preferable to check `shortCircuit` on each `next` function rather than just at the end of the chain. Then rather to mark each function call in the chain as "complete" or not. Then, the final check would just be that the final return value has `shortCircuit` or that the next called was a function that was itself marked as "complete". This then solves the problem of destructing of `next` that removes the `shortCircuit` property not correctly passing it on, which would only break when chained with a short circuiting loader and may not otherwise be caught during loader development.

@JakobJingleheimer let me know if I can clarify the above approach at all further.

Reply:
Thanks for catching!

Yep! I came to the same conclusion as you were outlining the scenario 😁 will update tonight and add a test case for it. For anyone else:

```js
export const resolve = async function resolveC(
  specifier,
  context,
  next,
) {
  const { url } = next(specifier);

  return {
    url: `${url}?foo=bar`
  };
}
```

correctly errors but cites the wrong loader:

```
Error [ERR_LOADER_CHAIN_INCOMPLETE]: The "resolve" hook from /…/loaderA.mjs did not call the next hook in its chain and did not explicitly signal a short circuit. If this is intentional, include `shortCircuit: true` in the hook's return.
```

Reply:
@JakobJingleheimer does the above need to be an error though?

Reply:
oh, actually no.